Announce `error-stack` (#628)
hashintel/hash: sites/hashdev/src/_pages/blog/4_announcement-error-stack.mdx

---

> authorPhoto: blog/4_announcement-error-stack/tim.jpg

Reviewer: ```suggestion
authorPhoto: blog/4_announcing-error-stack/tim.jpg
```

postPhoto: blog/2_open-source/OpenSource_Cover.jpg
title: "Announcing error-stack"
subtitle: "A context-aware error library with arbitrary attached user data"
author: "Tim Diekmann"

> jobTitle: "Backend Engineer at HASH"

Reviewer: ```suggestion
jobTitle: "Platform Engineer at HASH"
```
🥲

We haven't made it easy to create a new `Report` from a `String` because it requires the type to implement `Context`. It's still possible to use a newtype `UniqueError(String)` if a user wants to, but we discourage this usage because we've found that a lot of those places benefit from attaching a string message on a well-defined context instead. In any case, other libraries most probably still have to change the context from `UniqueError` to their context because of the explicit type requirement.

**By consciously imposing more restrictions in some areas we hope to reduce the _overall_ cost of development, in keeping with the philosophy underpinning Rust itself.**

> ## Some other cool things about `error-stack` that we don't have time for!

Reviewer: ```suggestion
### Some other cool things about `error-stack`
```

We say "we don't have space" below, and saying "we don't have time" could be interpreted the wrong way by someone only scanning the headings.

We don't have space here to talk about all of the features of `error-stack`, but here are a few more teasers. [Check out the repository](https://github.com/hashintel/hash/tree/main/packages/libs/error-stack) for even more information:

- The complete crate is written for `no-std` environments. However, when using `std`, a blanket implementation for `Context` for any `Error` is provided
- The blanket implementation for `Error` also makes the library compatible with almost all other libraries using the official trait
  - This means that if you pick this up, you should be able to migrate gradually (like we're doing with hEngine)
- Just like [`eyre`](https://crates.io/crates/eyre), the crate is able to set its own `Display` and `Debug` hooks that are called when printing the `Report`, which means you can write your own custom formatting and do other cool things
- We have in-built support for backtraces on the nightly channel and feature-flagged support for `tracing` to capture a `SpanTrace`

> ## The Future of `error-stack`

Reviewer: ```suggestion
### The future of `error-stack`
```
Future should not be capitalized.

---
authorPhoto: blog/4_announcement-error-stack/tim.jpg
postPhoto: blog/2_open-source/OpenSource_Cover.jpg
title: "Announcing error-stack"
subtitle: "A context-aware error library with arbitrary attached user data"
author: "Tim Diekmann"
jobTitle: "Backend Engineer at HASH"
date: "2022-06-10"
---

> # **Announcing `error-stack`**
> 

Reviewer 2: ```suggestion
```

The heading is already on the page (set by "title")

## Handling Errors across Module and Crate Boundaries

This led us into a pattern where we'd have a function called across a boundary (say from the `engine` module to the `simulation` crate), necessitating that we transform it between error types, and so for ergonomic reasons, we'd create a `From` implementation for the whole `Error` enumeration:

```rust
// mod engine

#[derive(ThisError)]
enum Error {
    #[error("Simulation error: {0}")]
    Simulation(#[from] crate::simulation::Error),
    ...
}
```

This would then often turn into error messages that looked like the following:

```
Error: Engine Error: Simulation Error: Experiment Error: Datastore Error: No such file or directory (os error 2)
```

> These error messages would tell us very little about the context that led to the error and weren't very helpful in actually tracking down the causes of problems. And while we have logging infrastructure in place, which helps with tracking down events that lead to errors, it relies on a few things:

Author: > helpful in [...] tracking down

> helps with tracking down

Can we come up with a better phrasing?

Reviewer 2: This phrasing is good imo, is there something specific you don't like about it?

I'm also good with "which is helpful in tracking down events that lead to errors"

Author: It's mainly because of using the same phrasing twice directly next to each other. I don't have a strong opinion, only wanted feedback from a native English speaker 🙂

Reviewer 2: Ah sorry I didn't actually spot the repeated phrase! Good call-out, although it's not too egregious here haha.

```suggestion
These error messages would tell us very little about the context that led to the error and weren't very helpful in actually tracking down the causes of problems. And while we have logging infrastructure in place, which is helpful with identifying events that lead to errors, it relies on a few things:
```

Author: Sounds good, thank you!

> These error messages would tell us very little about the context that led to the error and weren't very helpful in actually tracking down the causes of problems. And while we have logging infrastructure in place, which helps with tracking down events that lead to errors, it relies on a few things:

Author: > helpful in [...] tracking down

> helps with tracking down

Can we come up with a better phrasing?

Perhaps most importantly though, hEngine is a multi-threaded, multi-language project; this means that there will often be a mix of logs from various threads appearing before an error, with it being unclear as to which piece of logic caused the problem.

It's worth mentioning that `thiserror` is able to capture backtraces, which helps a fair bit. We started utilizing this by adding a `backtrace` field for some of our error variants, however, in doing so we ran into two issues.

Firstly, we needed to explicitly add the field to all error variants expected to have a backtrace:

```rust
// mod engine

#[derive(ThisError)]
enum Error {
    #[error("Simulation error: {0}")]
    Simulation(#[from] crate::simulation::Error, Backtrace),
    ...
}
```

Secondly, this required rerunning the program with `RUST_BACKTRACE=1`, or always running the program with it on.

Additionally, while backtraces help us locate the line of code which generated the error (which helps with disambiguating the example error message shown above), it's still missing a lot of other information that could prove helpful, like which iteration of a loop caused the error, or what path was missing.

> ## Defining Base Errors

Reviewer 2: ```suggestion
### Defining Base Errors
```

```rust
#[derive(ThisError)]
enum Error {
    /// Used when errors need to propagate but are too unique to be typed
    #[error("{0}")]
    Unique(String),
    ...
}

impl From<&str> for Error {
    fn from(s: &str) -> Self {
        Error::Unique(s.to_string())
    }
}

impl From<String> for Error {
    fn from(s: String) -> Self {
        Error::Unique(s)
    }
}
```

This was ergonomic at the time of error creation because we could quickly create one:

```rust
Error::from("Failed to initialize Context Package Creators")
```

However, in practice, we found ourselves often misusing these due to their relative convenience. This would lead to errors such as the following…

```rust
OrchestratorError::from(format!("Not a valid initial state file: {path:?}"))
```

…where we ended converting the `Path` to a (Debug) string rather than attaching the actual object in case needed at the point of handling.

This was through no fault of `thiserror`, but our usage of it in practice, and the patterns we found ourselves settling into.

> ## Error Conversions

Reviewer 2: ```suggestion
### Error Conversions
```

```rust
// mod experiment_controller
...
    .map_err(|experiment_err| Error::from(experiment_err.to_string()))?;
```

We already had an `experiment::Error` `From` `experiment_controller::Error` implementation, which meant that we were unable to define a `From` implementation going the other way. To get around this (initially temporarily, but as with many things, this ended up sticking around for a while) we simply turned the experiment error into a `String` and utilized the `Error::Unique` variant again.

This was clearly not ideal, and we ended up with a weird mixture of well-defined error variants and ones that were just captured in strings. The conversions also resulted in information loss depending on the `Display` implementation, including the loss of important pieces of information like the backtrace.

> # So, why **another Error Handling Library?**

Reviewer 2: ```suggestion
## So, why **another Error Handling Library?**
```

> We really love the Rust ecosystem, and although there are many great error crates out there, we struggled to find one that aligned with our needs, and which worked well with some of our project's inherent complexity. The closest error crate we could find was `[anyhow](https://crates.io/crates/anyhow)` (or its fork `[eyre](https://crates.io/crates/eyre)`), however, we had some additional wants:

Author: ```suggestion
We really love the Rust ecosystem, and although there are many great error crates out there, we struggled to find one that aligned with our needs, and which worked well with some of our project's inherent complexity. The closest error crate we could find was [`anyhow`](https://crates.io/crates/anyhow) (or its fork [`eyre`](https://crates.io/crates/eyre)), however, we had some additional wants:
```

> We really love the Rust ecosystem, and although there are many great error crates out there, we struggled to find one that aligned with our needs, and which worked well with some of our project's inherent complexity. The closest error crate we could find was `[anyhow](https://crates.io/crates/anyhow)` (or its fork `[eyre](https://crates.io/crates/eyre)`), however, we had some additional wants:

Author: ```suggestion
We really love the Rust ecosystem, and although there are many great error crates out there, we struggled to find one that aligned with our needs, and which worked well with some of our project's inherent complexity. The closest error crate we could find was [`anyhow`](https://crates.io/crates/anyhow) (or its fork [`eyre`](https://crates.io/crates/eyre)), however, we had some additional wants:
```

1.  encourage the user to provide a new error type if the scope is changed, usually by crossing module/crate boundaries (for example, a `ConfigParseError` when parsing a configuration file vs. an `IoError` when reading a file from disk)
2.  be able to use those error types within return types, without needing to handle difficult `From` logic
3.  be able to attach _any_ data to an error, not just string-like types, which then can be requested when handling the error.

Based on our experiences outlined above, we made a few additional design decisions that differentiated our approach to error handling from those existing crates designed to facilitate it. Generally speaking, these decisions were made to force developers to be more explicit when creating or chaining errors, but most importantly, we don't supply in-built support for creating errors from strings. While it is convenient, we want to encourage developers to think more about their errors by building concrete error types like `std::io::Error` or `std::num::ParseIntError`

> - _Note_: _It's still possible to create something like `Error::Unique` but we do not provide shortcut functionality for that within the library, hoping that a lot of its uses can instead be solved through “attachments”_

Author: ```suggestion
- _Note: It's still possible to create something like `Error::Unique` but we do not provide shortcut functionality for that within the library, hoping that a lot of its uses can instead be solved through “attachments”._
```

> - _Note_: _It's still possible to create something like `Error::Unique` but we do not provide shortcut functionality for that within the library, hoping that a lot of its uses can instead be solved through “attachments”_

Author: ```suggestion
- _Note: It's still possible to create something like `Error::Unique` but we do not provide shortcut functionality for that within the library, hoping that a lot of its uses can instead be solved through “attachments”._
```

> # What is **`error-stack`?**

Reviewer 2: ```suggestion
## What is **`error-stack`?**
```

> `error-stack` is centered around taking a base error and building up a full richer picture of it as it propagates. We call this a `Report` (you might recognize the naming choice of `Report` from the eyre crate, which we thought was fitting).

Author: ```suggestion
`error-stack` is centered around taking a base error and building up a full richer picture of it as it propagates. We call this a `Report` (you might recognize the naming choice of `Report` from the `eyre` crate, which we thought was fitting).
```

> `error-stack` is centered around taking a base error and building up a full richer picture of it as it propagates. We call this a `Report` (you might recognize the naming choice of `Report` from the eyre crate, which we thought was fitting).

Author: ```suggestion
`error-stack` is centered around taking a base error and building up a full richer picture of it as it propagates. We call this a `Report` (you might recognize the naming choice of `Report` from the `eyre` crate, which we thought was fitting).
```

    let path = path.as_ref();

    // First, we have a base error:
    let io_result = fs::read_to_string(path);      // Result<File, std::io::Error>

    // Convert the base error into a Report:
    let io_report = io_result.report();            // Result<File, Report<std::io::Error>>

    // Change the context, which will eventually return from the function:
    let config_report = io_report
        .change_context(ParseConfigError::new())?; // Result<File, Report<ParseConfigError>>

    // ...
}
```

Having the concept of a _current context_ has an important implication: when leaving the scope of a function it's likely that the return type from the caller function has a different type signature. This implies that a new Context _has to_ be added. This encourages the developer to think more about their error types and forces reasonable good cause traces.

For a `Context` we provide a trait, which is implemented for `std::error::Error` by default (if the `Error` is `Send + Sync + 'static`):

```rust
pub trait Context: Display + Debug + Send + Sync + 'static {
    fn provide<'a>(&'a self, demand: &mut Demand<'a>) {}
}
```

> `provide()` is an optional feature making use of the `[Provider` RFC](https://rust-lang.github.io/rfcs/3192-dyno.html?). When providing data here, they later can be requested by calling `Report::request_ref()` or `Report::request_value()`.

Author: ```suggestion
`provide()` is an optional feature making use of the [`Provider` RFC](https://rust-lang.github.io/rfcs/3192-dyno.html?). When providing data here, they later can be requested by calling `Report::request_ref()` or `Report::request_value()`.
```

> `provide()` is an optional feature making use of the `[Provider` RFC](https://rust-lang.github.io/rfcs/3192-dyno.html?). When providing data here, they later can be requested by calling `Report::request_ref()` or `Report::request_value()`.

Author: ```suggestion
`provide()` is an optional feature making use of the [`Provider` RFC](https://rust-lang.github.io/rfcs/3192-dyno.html?). When providing data here, they later can be requested by calling `Report::request_ref()` or `Report::request_value()`.
```

> **Note:** The implementation for the `Provider` RFC is not implemented at the time of writing, so we have vendored in the functionality for now. We will switch to the upstream implementation in `core::any`. **Using the `Provider` API currently requires the Rust nightly compiler.**

Author: ```suggestion
**Note:** The implementation for the `Provider` RFC was [merged just today](https://github.com/rust-lang/rust/pull/91970). We chose to not use the upstream implementation, for now, not to force users to update their compiler to the very last nightly, so we have vendored in the functionality for now. We will switch to the upstream implementation in `core::any` with version 0.2.0. **Using the `Provider` API currently requires the Rust nightly compiler.**
```

> **Note:** The implementation for the `Provider` RFC is not implemented at the time of writing, so we have vendored in the functionality for now. We will switch to the upstream implementation in `core::any`. **Using the `Provider` API currently requires the Rust nightly compiler.**

Author: ```suggestion
**Note:** The implementation for the `Provider` RFC was [merged just today](https://github.com/rust-lang/rust/pull/91970). We chose to not use the upstream implementation, for now, not to force users to update their compiler to the very last nightly, so we have vendored in the functionality for now. We will switch to the upstream implementation in `core::any` with version 0.2.0. **Using the `Provider` API currently requires the Rust nightly compiler.**
```

> ## Attachments

Reviewer 2: ```suggestion
### Attachments
```

Frequently you'll want to attach specific information to an error, requiring it to be encapsulated _within_ the error type, similar to crates like `anyhow` or `failure` where you are able to attach string-like types. However, `error-stack` is not only able to attach messages but also _any_ other data to the `Report`, for example, it's possible to attach a `Suggestion`. Using the example from above, we add a contextual message and a `Suggestion` (not offered by `error-stack`) to the `Report`:

```rust
fn parse_config(path: impl AsRef<Path>) -> Result<Config, Report<ParseConfigError>> {
    let path = path.as_ref();

    let content = fs::read_to_string(path)
        .report()
        .change_context(ParseConfigError::new())

>         .attach_lazy(|| format!("Could not read file {path:?}"))

Author: ```suggestion
        .attach_printable_lazy(|| format!("Could not read file {path:?}"))
```

>         .attach_lazy(|| format!("Could not read file {path:?}"))

Author: ```suggestion
        .attach_printable_lazy(|| format!("Could not read file {path:?}"))
```

        .attach(Suggestion("Use a file you can read next time!"))?;

    // ...
}
```

It's then possible to request the `Suggestion`s from the `Report`, which returns an iterator (most recent attachment first) as you can attach the same type multiple times:

```rust
if let Err(report) = parse_config("config.json") {
    eprintln!("{report:?}");
    for suggestion in report.request_ref::<Suggestion>() {
        eprintln!("Suggestion: {suggestion}");
    }
}
```

**Note:** As mentioned before, calling `request_ref()` on a `Report` **currently requires a nightly compiler** as this is using the Provider API.

This will eventually output:

```
Could not parse configuration file
             at main.rs:6:10

>       - Use a file you can read next time!
>       - Could not read file "config.json"

Author: ```suggestion
      - Could not read file "config.json"
      - 1 additional opaque attachment
```

>       - Use a file you can read next time!
>       - Could not read file "config.json"

Author: ```suggestion
      - Could not read file "config.json"
      - 1 additional opaque attachment
```

Caused by:
   0: No such file or directory (os error 2)
             at main.rs:5:10

Stack backtrace:
   0: error_stack::report::Report<T>::new
             at error-stack/src/report.rs:187:18
   1: error_stack::context::<impl core::convert::From<C> for error::report::Report<C>>::from
             at error-stack/src/context.rs:87:9
   2: <core::result::Result<T,E> as error::result::IntoReport>::report
             at error-stack/src/result.rs:204:31
   3: parse_config
             at main.rs:4:19
   4: main
             at main.rs:13:26
   5: ...

Suggestion: Use a file you can read next time!
```

> ## What's the catch?

Reviewer 2: ```suggestion
### What's the catch?
```

You might think that this comes with a lot of performance overhead, but with speed being a key concern of hEngine, we've taken this into account. The impact on the ‘happy' success code path is minimal; a `Report` is only one pointer in size and requesting from a `Report` only happens at the time of error reporting. We use a similar data layout as `anyhow` and thus expect similar performance.

> However, we've purposefully added in some requirements that result in a little bit of extra upfront **development** overhead. We think having explicitly typed `Result` values is worth requiring, and especially helpful in identifying different types across module/crate boundaries, so we still want users to create types (`Context`s). This approach makes `error-stack` usable not only for binaries but also for libraries. In most cases, the caller of a library only cares about the error directly returned from a library, not an underlying OS error. As the current context is _always_ known, the user directly has all type information without optimistically trying to downcast to an error type (which remains possible). This also implies that more times than not a developer is _forced_ to add a new context because the type system requires it. We've found from experience that this automatically greatly improves error reporting.

Author: ```suggestion
However, we've purposefully added in some requirements that result in a little bit of extra upfront **development** overhead. We think having explicitly typed `Result` values is worth requiring, and especially helpful in identifying different types across module/crate boundaries, so we still want users to create types (`Context`s). This approach makes `error-stack` usable not only for binaries but also for libraries. In most cases, the caller of a library only cares about the error directly returned from a library, not an underlying OS error. As the current context is _always_ known, the user directly has all type information without optimistically trying to downcast to an error type (which remains possible). This also implies that **more times than not** a developer is _forced_ to add a new context because the type system requires it. We've found from experience that this automatically greatly improves error reporting.
```

> However, we've purposefully added in some requirements that result in a little bit of extra upfront **development** overhead. We think having explicitly typed `Result` values is worth requiring, and especially helpful in identifying different types across module/crate boundaries, so we still want users to create types (`Context`s). This approach makes `error-stack` usable not only for binaries but also for libraries. In most cases, the caller of a library only cares about the error directly returned from a library, not an underlying OS error. As the current context is _always_ known, the user directly has all type information without optimistically trying to downcast to an error type (which remains possible). This also implies that more times than not a developer is _forced_ to add a new context because the type system requires it. We've found from experience that this automatically greatly improves error reporting.

Author: ```suggestion
However, we've purposefully added in some requirements that result in a little bit of extra upfront **development** overhead. We think having explicitly typed `Result` values is worth requiring, and especially helpful in identifying different types across module/crate boundaries, so we still want users to create types (`Context`s). This approach makes `error-stack` usable not only for binaries but also for libraries. In most cases, the caller of a library only cares about the error directly returned from a library, not an underlying OS error. As the current context is _always_ known, the user directly has all type information without optimistically trying to downcast to an error type (which remains possible). This also implies that **more times than not** a developer is _forced_ to add a new context because the type system requires it. We've found from experience that this automatically greatly improves error reporting.
```

We haven't made it easy to create a new `Report` from a `String` because it requires the type to implement `Context`. It's still possible to use a newtype `UniqueError(String)` if a user wants to, but we discourage this usage because we've found that a lot of those places benefit from attaching a string message on a well-defined context instead. In any case, other libraries most probably still have to change the context from `UniqueError` to their context because of the explicit type requirement.

**By consciously imposing more restrictions in some areas we hope to reduce the _overall_ cost of development, in keeping with the philosophy underpinning Rust itself.**

## Some other cool things about `error-stack` that we don't have time for!

> We don't have space here to talk about all of the features of `error-stack`, but here are a few more teasers. Check out the repository for even more information:

Author: ```suggestion
We don't have space here to talk about all of the features of `error-stack`, but here are a few more teasers. [Check out the repository](https://github.com/hashintel/hash/tree/main/packages/libs/error-stack) for even more information:
```

> We don't have space here to talk about all of the features of `error-stack`, but here are a few more teasers. Check out the repository for even more information:

Author: ```suggestion
We don't have space here to talk about all of the features of `error-stack`, but here are a few more teasers. [Check out the repository](https://github.com/hashintel/hash/tree/main/packages/libs/error-stack) for even more information:
```

> ## Some other cool things about `error-stack` that we don't have time for!

Reviewer 2: ```suggestion
### Some other cool things about `error-stack` that we don't have time for!
```

Reviewer: @Alfred-Mountfield Closing this in favor of my suggested alteration to this header (https://github.com/hashintel/hash/pull/628/files#r894362708). It incorporates your suggestion here.

- The complete crate is written for `no-std` environments. However, when using `std`, a blanket implementation for `Context` for any `Error` is provided
- The blanket implementation for `Error` also makes the library compatible with almost all other libraries using the official trait
  - This means that if you pick this up, you should be able to migrate gradually (like we're doing with hEngine)

> - Just like `[eyre](https://crates.io/crates/eyre)`, the crate is able to set its own `Display` and `Debug` hooks that are called when printing the `Report`, which means you can write your own custom formatting and do other cool things

Author: ```suggestion
- Just like [`eyre`](https://crates.io/crates/eyre), the crate is able to set its own `Display` and `Debug` hooks that are called when printing the `Report`, which means you can write your own custom formatting and do other cool things
```

> - Just like `[eyre](https://crates.io/crates/eyre)`, the crate is able to set its own `Display` and `Debug` hooks that are called when printing the `Report`, which means you can write your own custom formatting and do other cool things

Author: ```suggestion
- Just like `[eyre](https://crates.io/crates/eyre)`, the crate is able to set its own `Display` and `Debug` hooks that are called when printing the `Report`, which means you can write your own custom formatting and do other cool things
```
```suggestion
- Just like [`eyre`](https://crates.io/crates/eyre), the crate is able to set its own `Display` and `Debug` hooks that are called when printing the `Report`, which means you can write your own custom formatting and do other cool things
```

> - We have in-built support for backtraces on the nightly channel and feature-flagged support for `tracing` and capturing a `SpanTrace`

Author: ```suggestion
- We have in-built support for backtraces on the nightly channel and feature-flagged support for `tracing` to capture a `SpanTrace`
```

> - We have in-built support for backtraces on the nightly channel and feature-flagged support for `tracing` and capturing a `SpanTrace`

Author: ```suggestion
- We have in-built support for backtraces on the nightly channel and feature-flagged support for `tracing` to capture a `SpanTrace`
```

> ## The Future of `error-stack`

Reviewer 2: ```suggestion
### The Future of `error-stack`
```

Reviewer: Closing this suggestion in favor of https://github.com/hashintel/hash/pull/628/files#r894363610

As previously noted, we are still in the process of migrating our own codebase to `error-stack` and we intend to make further improvements as we apply the crate to more of hEngine and our internal code at [HASH](https://hash.ai/).

> We will switch to the provider implementation in `core::any` as soon as it lands (the main reason why this API currently requires the nightly release channel).

Author: ```suggestion
We will switch to the provider implementation in `core::any` in version 0.2.0 (the main reason why this API currently requires the nightly release channel).
```

> We will switch to the provider implementation in `core::any` as soon as it lands (the main reason why this API currently requires the nightly release channel).

Author: ```suggestion
We will switch to the provider implementation in `core::any` in version 0.2.0 (the main reason why this API currently requires the nightly release channel).
```

We cannot guarantee full backward compatibility until this crate hits `v1.0.0`. However, we will aim to avoid breaking changes, and in the event of any will announce them beforehand and provide a migration guide.

Contributions are greatly welcome. View the [project code on GitHub](https://github.com/hashintel/hash/tree/main/packages/libs/error-stack) and, if you like it, please star the repository!

> ## Acknowledgments

Reviewer 2: ```suggestion
### Acknowledgments
```

We would like to thank the [Error Project Group](https://github.com/rust-lang/project-error-handling) for their work on the `Error` trait and the [Provider RFC](https://rust-lang.github.io/rfcs/3192-dyno.html). It's unlikely we would have written this library without the latter — or at least not in a way we would have been truly satisfied with.

> ## License

Reviewer 2: ```suggestion
### License
```
